ratelimit_test.go

```diff
@@ -13,6 +13,82 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+type runner interface {
+	// startTaking tries to Take() on passed in limiters in a loop/goroutine.
+	startTaking(rls ...ratelimit.Limiter)
+	// assertCountAt asserts the limiters have Taken() a number of times at the given time.
+	// It's a thin wrapper around afterFunc to reduce boilerplate code.
+	assertCountAt(d time.Duration, count int)
+	// getClock returns the test clock.
+	getClock() ratelimit.Clock
+	// afterFunc executes a func at a given time.
+	afterFunc(d time.Duration, fn func())
+}
+
+type runnerImpl struct {
+	t *testing.T
+
+	clock *clock.Mock
+	count atomic.Int32
+	// maxDuration is the time we need to move into the future for a test.
+	// It's populated automatically based on assertCountAt/afterFunc.
+	maxDuration time.Duration
+	doneCh      chan struct{}
+	wg          sync.WaitGroup
+}
+
+func runTest(t *testing.T, fn func(runner)) {
+	r := runnerImpl{
+		t:      t,
+		clock:  clock.NewMock(),
+		doneCh: make(chan struct{}),
+	}
+	defer close(r.doneCh)
+	defer r.wg.Wait()
+
+	fn(&r)
+	r.clock.Add(r.maxDuration)
+}
+
+// startTaking tries to Take() on passed in limiters in a loop/goroutine.
+func (r *runnerImpl) startTaking(rls ...ratelimit.Limiter) {
+	go func() {
```

Comment:
do we need to add this to the waitgroup?

Reply:
So, technically, I think we should **however** it's not needed right now - `clock.afterFunc` does `time.sleep(1*millisecond)` so we get it for free.

Given that it works now, and the old code was doing this, I'd prefer to keep it as is - just to minimize changes in this PR.

#39 actually adds a separate wg for starting jobs (it's out of date now, I'll update it after this PR) - I'd prefer to iterate & update it in follow ups.

But yeah, more wgs will be needed to deprecate the internal clock for.